arches/app/permissions/arches_standard.py

```diff
@@ -0,0 +1,926 @@
+                    )
+                    results["hits"]["hits"] += results_scrolled["hits"]["hits"]
+            restricted_ids = [res["_id"] for res in results["hits"]["hits"]]
+            return restricted_ids
+
+    def update_groups_for_user(self, user: User) -> None:
+        """Hook for spotting group updates on a user."""
+        ...
+
+    def update_permissions_for_user(self, user: User) -> None:
+        """Hook for spotting permission updates on a user."""
+        ...
+
+    def update_permissions_for_group(self, group: Group) -> None:
+        """Hook for spotting permission updates on a group."""
+        ...
+
+    def user_has_resource_model_permissions(
+        self,
+        user: User,
+        perms: str | Iterable[str],
+        resource: ResourceInstance | None = None,
+        graph_id: str | None = None,
+    ) -> bool:
+        """
+        Checks if a user has any explicit permissions to a model's nodegroups
+
+        Arguments:
+        user -- the user to check
+        perms -- the permssion string eg: "read_nodegroup" or list of strings
+        graph_id -- a graph id to check if a user has permissions to that graph's type specifically
+
+        """
+
+        if resource:
+            graph_id = resource.graph_id
+
+        nodegroups = self.get_nodegroups_by_perm(user, perms)
+        nodes = (
+            Node.objects.filter(nodegroup__in=nodegroups)
+            .filter(graph_id=graph_id)
+            .select_related("graph")
+        )
+        return bool(nodes.exists())
+
+    def user_can_read_resource(
+        self, user: User, resourceid: str | None = None
+    ) -> bool | None:
+        """
+        Requires that a user be able to read an instance and read a single nodegroup of a resource
+
+        """
+        if user.is_authenticated:
+            if user.is_superuser:
+                return True
+            if resourceid is not None and resourceid != "":
+                result = self.check_resource_instance_permissions(
+                    user, resourceid, "view_resourceinstance"
+                )
+                if result is not None:
+                    if result["permitted"] == "unknown":
+                        return self.user_has_resource_model_permissions(
+                            user, ["models.read_nodegroup"], result["resource"]
+                        )
+                    else:
+                        return result["permitted"]
+                else:
+                    return None
+
+            return (
+                len(self.get_resource_types_by_perm(user, ["models.read_nodegroup"]))
+                > 0
+            )
+        return False
+
+    def get_resource_types_by_perm(
+        self, user: User, perms: str | Iterable[str]
+    ) -> list[str]:
+        graphs = set()
+        nodegroups = self.get_nodegroups_by_perm(user, perms)
+        for node in Node.objects.filter(nodegroup__in=nodegroups).prefetch_related(
```

Comment:
select_related when going from many -> one, otherwise an unnecessary query is made.
```suggestion
        for node in Node.objects.filter(nodegroup__in=nodegroups).select_related(
```